### PR TITLE
HBASE-23066 Allow cache on write during compactions when prefetching …

### PR DESCRIPTION
**Pull request by jacob-leblanc**
…is enabled

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 37s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :green_heart: |  mvninstall  |   5m 35s |  master passed  |
| :green_heart: |  compile  |   0m 56s |  master passed  |
| :green_heart: |  checkstyle  |   1m 20s |  master passed  |
| :green_heart: |  shadedjars  |   4m 34s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 38s |  master passed  |
| :blue_heart: |  spotbugs  |   4m  3s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m  0s |  master passed  |
||| _ Patch Compile Tests _ |
| :green_heart: |  mvninstall  |   4m 49s |  the patch passed  |
| :green_heart: |  compile  |   0m 56s |  the patch passed  |
| :green_heart: |  javac  |   0m 56s |  the patch passed  |
| :green_heart: |  checkstyle  |   1m 20s |  the patch passed  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 34s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m 48s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 36s |  the patch passed  |
| :green_heart: |  findbugs  |   4m  3s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  | 159m 57s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 35s |  The patch does not generate ASF License warnings.  |
|  |   | 216m 42s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-707/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/707 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux fb96839beab8 4.15.0-60-generic #67-Ubuntu SMP Thu Aug 22 16:55:30 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-707/out/precommit/personality/provided.sh |
| git revision | master / ba12d5b9de |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-707/1/testReport/ |
| Max. process+thread count | 5023 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-707/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by ramkrish86** [comment]
LGTM. Let's wait for some time for others to review.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CacheConfig.java [hunk 1]
```diff
@@ -81,6 +81,13 @@
    */
   public static final String PREFETCH_BLOCKS_ON_OPEN_KEY = "hbase.rs.prefetchblocksonopen";
 
+  /**
+   * Configuration key to cache blocks when a compacted file is written, predicated on prefetching
+   * being enabled for the column family.
+   */
+  public static final String PREFETCH_COMPACTED_BLOCKS_ON_WRITE_KEY =
```
Comment (anoopsjohn) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CacheConfig.java]: A bit confusing.. Are we doing the prefetch of the new compacted file once it is written?
Dont think so.. When we write the file, that time itself the caching happens. So it is cache on write. Why its called prefetch then?  There is no extra fetch op happening right?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CacheConfig.java [hunk 2]
```diff
@@ -319,6 +337,13 @@ public boolean shouldPrefetchOnOpen() {
     return this.prefetchOnOpen;
   }
 
+  /**
+   * @return true if blocks should be cached while writing during compaction, false if not
+   */
+  public boolean shouldCacheCompactedBlocksOnWrite() {
+    return this.prefetchCompactedDataOnWrite && this.prefetchOnOpen;
```
Comment (anoopsjohn) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CacheConfig.java]: Oh.. So the cache on write (at compaction) happens iff prefetch config is ON ! Anyways in ur case the prefetch which is another config, is ON right?  I think this is the reason why the new config you have named that way.  But some how I feel that config name is bit misleading.

Comment (anoopsjohn) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CacheConfig.java]: Actually the cache size should be much bigger than the hot data set size if u want to do cache on compact.  Because the compacted away data might be already in cache (Those are flused files or a result of another compaction). Those are recently been accessed also (by the compaction thread).  This feature should be very carefully used.

Comment (jacob-leblanc) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/CacheConfig.java]: Thanks for looking at this. My understanding is that in cases where prefetch is enabled, the new file is going to be read into the cache after compaction completes anyway. So the cache size requirements are the same when this new setting is enabled. This is why I wanted to limit the scope of the cache on write to only apply where prefetching is enabled: it simply is a way to do the cache loading more efficiently while we are writing the data out rather than having to read it back after compaction is done which I've found is very expensive when data is in S3.

As far as the name goes, I struggled to come up with something intuitive - how do I explain in the name alone that this only applies when prefetching is on? I tried to convey "when prefetching, do the prefetch of compacted data on write." I'm not in love with the name and I'm open to suggestions. I didn't want to give the false impression that all compacted data is going to be cached on write. Maybe "cacheCompactedDataOnWriteIfPrefetching"? Is that too wordy?

**Comment by busbey** [comment]
is this made obsolete by #919?

**Comment by virajjasani** [comment]
> is this made obsolete by #919?

I believe so. Closing this PR @jacob-leblanc @ramkrish86
